### PR TITLE
Add a busy event counter

### PR DESCRIPTION
**Pull request by DSmolonogov**
# Description

Add a busy event counter 

Fixes # (issue)

### CODE REVIEW

File: pipeline/event.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package pipeline
 
 import (
 	"fmt"
+	"github.com/ozontech/file.d/stats"
```
Comment (vadimalekseev) [inline pipeline/event.go:5]: Setup goimports, please

File: pipeline/event.go [hunk 2]
```diff
@@ -208,7 +211,7 @@ type eventPool struct {
 	getCond *sync.Cond
 }
 
-func newEventPool(capacity int) *eventPool {
+func newEventPool(capacity int, pipelineName string) *eventPool {
```
Comment (vadimalekseev) [inline pipeline/event.go:214]: You forgot to set the value in the struct

File: pipeline/pipeline.go [hunk 3]
```diff
@@ -581,7 +587,7 @@ func (p *Pipeline) logChanges(myDeltas *deltas) {
 	p.logger.Infof(`%q pipeline stats interval=%ds, active procs=%d/%d, queue=%d/%d, out=%d|%.1fMb,`+
 		`rate=%d/s|%.1fMb/s, read ops=%d/s, total=%d|%.1fMb, avg size=%d, max size=%d, pool fullness=%d/%d`,
 		p.Name, interval/time.Second, p.activeProcs.Load(), p.procCount.Load(),
-		p.settings.Capacity-p.eventPool.freeEventsCount, p.settings.Capacity,
+		p.eventPool.workEventsCount, p.settings.Capacity,
 		int64(myDeltas.deltaInputEvents), float64(myDeltas.deltaInputSize)/1024.0/1024.0, rate, rateMb, readOps,
 		inputEvents, float64(inputSize)/1024.0/1024.0, inputSize/tc, p.maxSize,
 		p.eventPool.capacity-p.eventPool.freeEventsCount, p.eventPool.capacity)
```
Comment (vadimalekseev) [inline pipeline/pipeline.go:593]: Is `p.eventPool.capacity-p.eventPool.freeEventsCount ` redundant here?

File: pipeline/event.go [hunk 4]
```diff
@@ -195,9 +196,11 @@ func (e *Event) String() string {
 
 // channels are slower than this implementation by ~20%
 type eventPool struct {
-	capacity int
+	capacity     int
+	pipelineName string
 
 	freeEventsCount int
```
Comment (vadimalekseev) [inline pipeline/event.go:202]: Is `freeEventsCount` used?

File: pipeline/event.go [hunk 5]
```diff
@@ -260,7 +263,8 @@ func (p *eventPool) get() *Event {
 	event := p.events[x]
 	p.events[x] = nil
 	p.free2[x].Store(false)
-
+	p.workEventsCount.Inc()
+	stats.GetGauge(p.pipelineName, workEventsGauge).Inc()
```
Comment (vadimalekseev) [inline pipeline/event.go:267]: It will not work if implementation of the `Pipeline.subsystem()` is changed

**Review by vadimalekseev — CHANGES_REQUESTED** [review]

File: pipeline/event.go (hunk 5 above)
Comment (vadimalekseev) [reply pipeline/event.go:267]: Let's move the metric registration to the event pool structure

File: pipeline/pipeline.go (hunk 3 above)
Comment (DSmolonogov) [reply pipeline/pipeline.go:593]: It is redundant. Pool fullness contains the same information as queue. 
Сan i delete it?

Comment (vadimalekseev) [reply pipeline/pipeline.go]: Yep :)

File: pipeline/pipeline.go [hunk 6]
```diff
@@ -38,6 +39,8 @@ const (
 	antispamUnbanIterations = 4
 	metricsGenInterval      = time.Hour
 
+	workEventsGauge         = "work_events_gauge"
```
Comment (vitkovskii) [inline pipeline/pipeline.go:42]: https://prometheus.io/docs/practices/naming/, don't add metric type into metric name

Comment (vitkovskii) [reply pipeline/pipeline.go:42]: better report pool size and events in pool rather than work/wait event count

File: pipeline/pipeline.go [hunk 7]
```diff
@@ -578,13 +589,29 @@ func (p *Pipeline) logChanges(myDeltas *deltas) {
...
-		`rate=%d/s|%.1fMb/s, read ops=%d/s, total=%d|%.1fMb, avg size=%d, max size=%d, pool fullness=%d/%d`,
+	workEventValue, err := p.GetValueGauge(p.eventPool.subsystemName(), workEventsGauge)
+	if err != nil {
+		p.logger.Infof("failed to get workEventValue: " + err.Error())
+	}
+	waitEventValue, err := p.GetValueGauge(p.subsystemName(), waitEventsGauge)
+	if err != nil {
+		p.logger.Infof("failed to get waitEventValue: " + err.Error())
+	}
+	p.logger.Infof(`%q pipeline stats interval=%ds, active procs=%d/%d, working events=%d/%d, in queue=%d, out=%d|%.1fMb,`+
```
Comment (vitkovskii) [inline pipeline/pipeline.go:600]: let's add only log, `events in pool=%d/%d`, `events outsize pool=%d`

File: pipeline/pipeline.go [hunk 8]
```diff
@@ -578,13 +589,29 @@ func (p *Pipeline) logChanges(myDeltas *deltas) {
 	readOps := int(myDeltas.deltaReads * float64(time.Second) / float64(interval))
 	tc := int64(math.Max(float64(inputSize), 1))
 
-	p.logger.Infof(`%q pipeline stats interval=%ds, active procs=%d/%d, queue=%d/%d, out=%d|%.1fMb,`+
-		`rate=%d/s|%.1fMb/s, read ops=%d/s, total=%d|%.1fMb, avg size=%d, max size=%d, pool fullness=%d/%d`,
+	workEventValue, err := p.GetValueGauge(p.eventPool.subsystemName(), workEventsGauge)
+	if err != nil {
+		p.logger.Infof("failed to get workEventValue: " + err.Error())
```
Comment (vadimalekseev) [inline pipeline/pipeline.go:594]: Use `failed to get workEventValue: %s` instead of concatenation

File: pipeline/event.go [hunk 9]
```diff
@@ -260,7 +263,8 @@ func (p *eventPool) get() *Event {
 	event := p.events[x]
 	p.events[x] = nil
 	p.free2[x].Store(false)
-
+	metric.GetGauge(p.subsystemName(), workEventsGauge).Inc()
```
Comment (vano144) [inline pipeline/event.go:266]: Can we call metric.GetGauge only once and inc here result of  metric.GetGauge(p.subsystemName(), workEventsGauge) ?

Comment (DSmolonogov) [reply pipeline/event.go:266]: We make a choice, either add two counters, and conveniently write metrics to the logs, or leave only the prometheus metric and get the value in the logs through error handling.

File: pipeline/event_test.go [hunk 10]
```diff
@@ -20,8 +24,9 @@ func TestEventPoolDump(t *testing.T) {
 
 func BenchmarkEventPoolOneGoroutine(b *testing.B) {
 	const capacity = 32
+	metric.InitStats()
 
-	p := newEventPool(capacity, DefaultAvgInputEventSize)
+	p := newEventPool(2, DefaultAvgInputEventSize, "")
```
Comment (vadimalekseev) [inline pipeline/event_test.go:29]: Why `2` instead of `capacity`?

File: pipeline/event.go [hunk 11]
```diff
@@ -193,29 +195,33 @@ func (e *Event) String() string {
...
-	getCounter      atomic.Int64
-	backCounter     atomic.Int64
-	events          []*Event
-	free1           []atomic.Bool
-	free2           []atomic.Bool
+	capacity     int
+	pipelineName string
+
+	avgEventSize      int
+	WorkEventPoolProm prometheus.Gauge
```
Comment (vadimalekseev) [inline pipeline/event.go:202]: Should we export `WorkEventPoolProm`?

File: pipeline/pipeline.go [hunk 12]
```diff
@@ -185,6 +189,16 @@ func (p *Pipeline) subsystemName() string {
 }
 
 func (p *Pipeline) registerMetrics() {
+	metric.RegisterGauge(&metric.MetricDesc{
+		Subsystem: p.subsystemName(),
+		Name:      inUseEventsGauge,
+		Help:      "Running events counter",
```
Comment (vitkovskii) [inline pipeline/pipeline.go:195]: `Count of pool events which is used for processing`

File: pipeline/pipeline.go [hunk 13]
```diff
@@ -37,12 +38,14 @@ const (
 	antispamUnbanIterations = 4
 	metricsGenInterval      = time.Hour
 
+	inUseEventsGauge        = "in_use_events"
```
Comment (vitkovskii) [inline pipeline/pipeline.go:41]: `event_pool_in_use_events`

Comment (vitkovskii) [reply pipeline/pipeline.go:41]: and let's move closer to `event_pool_capacity` metric

File: pipeline/pipeline.go [hunk 14]
```diff
@@ -570,20 +589,20 @@ type deltas struct {
 func (p *Pipeline) logChanges(myDeltas *deltas) {
 	inputSize := p.inputSize.Load()
 	inputEvents := p.inputEvents.Load()
+	workEventValue := p.eventPool.inUseEvents.Load()
```
Comment (vitkovskii) [inline pipeline/pipeline.go:592]: `workEventValue` => `inUseEvents`

**Review by vitkovskii — CHANGES_REQUESTED** [review]

**Review by vitkovskii — APPROVED** [review]

**Review by vadimalekseev — APPROVED** [review]

**Pull request merged**